test/integration/consul-container/test/envoy_extensions/wasm_test.go

```diff
@@ -0,0 +1,519 @@
+	}
+
+	err = buildC.Start(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// The testcontainers.ContainerFile does not honor the filemode set on the files for some reason
+	// so we can just chown them to the default user on the image
+	_, _, err = buildC.Exec(ctx, []string{
+		"sudo",
+		"chown",
+		"tinygo",
+		fmt.Sprintf("%s/wasm_add_header.go", containerDir),
+		fmt.Sprintf("%s/go.mod", containerDir),
+		fmt.Sprintf("%s/go.sum", containerDir),
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	_, _, err = buildC.Exec(ctx, []string{
+		"tinygo",
+		"build",
+		"-o",
+		containerWASMFile,
+		"-scheduler=none",
+		"-target=wasi",
+		fmt.Sprintf("%s/wasm_add_header.go", containerDir),
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	r, err := buildC.CopyFileFromContainer(ctx, containerWASMFile)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer r.Close()
+
+	f, err := os.Create(fmt.Sprintf("%s/wasm_add_header.go.wasm", hostWASMDir))
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer f.Close()
+
+	_, err = io.Copy(f, r)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	defer buildC.Terminate(ctx)
+}
+
+func buildNginxFileServer(t *testing.T, conf testcontainers.ContainerFile, files ...testcontainers.ContainerFile) string {
+	req := testcontainers.ContainerRequest{
+		// nginx:stable
+		Image:        "nginx@sha256:b07a5ab5292bd90c4271a55a44761899cc1b14814172cf7f186e3afb8bdbec28",
+		Name:         "nginx-fileserver",
+		ExposedPorts: []string{"80/tcp"},
+		WaitingFor:   wait.ForHTTP("/"),
+		LifecycleHooks: []testcontainers.ContainerLifecycleHooks{
+			{
+				PostStarts: []testcontainers.ContainerHook{
+					func(ctx context.Context, c testcontainers.Container) error {
+						_, _, err := c.Exec(ctx, []string{"mkdir", "-p", "/www/downloads"})
+						if err != nil {
+							return err
+						}
+
+						for _, f := range files {
+							fBytes, err := os.ReadFile(f.HostFilePath)
+							if err != nil {
+								return err
+							}
+							err = c.CopyToContainer(ctx, fBytes, f.ContainerFilePath, f.FileMode)
+							if err != nil {
+								return err
+							}
+
+							_, _, err = c.Exec(ctx, []string{"chmod", "+r", f.ContainerFilePath})
+							if err != nil {
+								return err
+							}
+						}
+
+						return err
+					},
+				},
+			},
+		},
+		Files: []testcontainers.ContainerFile{conf},
+	}
+
+	ctx := context.Background()
+
+	nginxC, err := testcontainers.GenericContainer(ctx, testcontainers.GenericContainerRequest{
+		ContainerRequest: req,
+		Started:          false,
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = nginxC.Start(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	ip, err := nginxC.Host(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	mappedPort, err := nginxC.MappedPort(ctx, "80")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	return fmt.Sprintf("http://%s:%s", ip, mappedPort.Port())
+}
+
+func chain(fns ...func(testcontainers.ContainerRequest) testcontainers.ContainerRequest) func(testcontainers.ContainerRequest) testcontainers.ContainerRequest {
```

Comment:
Agreed. I'm not super happy with the `chain` bit here but it felt like it made the function signatures more clear than variadic functions (plus a few of them were already variadic for other bits of configuration).

The downside is that in the functions that actually call these configuration functions we have to check that the param is not `nil` rather than looping over the range of input functions.

Reply:
See https://github.com/hashicorp/consul/pull/17756/files#diff-e60cd58fca8c14da51ee363a394f2306cfdf11554329893af09da72761b5c31dR261

Reply:
@cthain Not sure where the best spot to put these is. Should we leave them here until we have a need for them in multiple spots?

Reply:
Yes I think that makes sense. It may be that we only need this here for now so there's no sense in over-engineering it at this point.